design/src/rfcs/rfc0031_implementing_adaptive_retry_behavior.md

```diff
@@ -0,0 +1,162 @@
+# Implementing Adaptive Retry Behavior
+
+> Status: RFC
+>
+> Applies to: client, server
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+This RFC defines how the Rust SDK will implement Adaptive Retry Behavior. This new behavior works on the principle of client-side rate limiting and only takes effect once a client has received a throttling error. The new behavior:
+- Is opt-in. The existing retry behavior will remain the default.
+- Will perform similarly to the existing retry behavior when comparing the rate of successful requests.
+- Will improve the experience of customers that use the SDK to send very large numbers of concurrent requests.
+
```

Comment:
I can see that the RFC is a draft but I had a few comments so I thought I might as well add them now.

1) I couldn't find anything on how does one turn off retries. Is it going to be:

```
    let retry_config = RetryConfig::disabled();
```

Which I am assuming is the same as setting:

```
    let retry_config = RetryConfig::standard().with_max_attempts(1);
```

2) Are we going to allow retry policy to be configured separately for specific operations? If yes, what does the API look for that?

3) The RFC doesn't mention the [RetryClassifier](https://github.com/awslabs/smithy-rs/blob/e2e0d9354548bbf2d76b660571cd0701d016771b/rust-runtime/aws-smithy-client/src/retry.rs#L375) that each operation can have. Can we please add a section on how does that play a role in the policy behavior? and, how does one specify a RetryClassifier on a particular operation or the overall client.

4) What metrics will the client emit by default? and:
a) Can a user disable metrics? 
b) Can a custom metric be emitted on a particular lifecycle event?